File: frontend/src/lib/api/access-list.ts
import { apiClient } from "./client";
import { VyOSResponse } from "../types/api";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface AccessListRule {
  rule_number: number;
  action: string;  // permit or deny
  description?: string | null;
  source_type?: string | null;  // any, host, inverse-mask, network
  source_address?: string | null;
  source_mask?: string | null;
  source_exact_match?: boolean;  // IPv6 exact-match flag (can coexist with any)
  destination_type?: string | null;  // any, host, inverse-mask, network
  destination_address?: string | null;
  destination_mask?: string | null;
}

export interface AccessList {
  number: string;  // For IPv4, name for IPv6
  description?: string | null;
  rules: AccessListRule[];
  list_type: string;  // ipv4 or ipv6
}

export interface AccessListConfigResponse {
  ipv4_lists: AccessList[];
  ipv6_lists: AccessList[];
  total_ipv4: number;
  total_ipv6: number;
}

export interface AccessListCapabilitiesResponse {
  version: string;
  features: {
    ipv4_access_list: {
      supported: boolean;
      description: string;
    };
    ipv6_access_list: {
      supported: boolean;
      description: string;
    };
    source_destination_filters: {
      supported: boolean;
      description: string;
    };
  };
  access_list_ranges: {
    standard: string;
    extended: string;
  };
  device_name?: string;
}

export interface AccessListBatchOperation {
  op: string;
  value?: string;
  value2?: string;
}

export interface AccessListBatchRequest {
  identifier: string;  // number for IPv4, name for IPv6
  list_type: string;  // ipv4 or ipv6
  rule_number?: number;
  operations: AccessListBatchOperation[];
}

// ============================================================================
// API Service
// ============================================================================

class AccessListService {
  /**
   * Get capabilities based on VyOS version
   */
  async getCapabilities(): Promise<AccessListCapabilitiesResponse> {
    return apiClient.get<AccessListCapabilitiesResponse>("/vyos/access-list/capabilities");
  }

  /**
   * Get all access-list configurations
   */
  async getConfig(refresh: boolean = false): Promise<AccessListConfigResponse> {
    return apiClient.get<AccessListConfigResponse>("/vyos/access-list/config", {
      refresh: refresh.toString(),
    });
  }

  /**
   * Refresh the cached configuration
   */
  async refreshConfig(): Promise<VyOSResponse> {
    return apiClient.post<VyOSResponse>("/vyos/config/refresh");
  }

  /**
   * Execute batch operations
   */
  async batchConfigure(request: AccessListBatchRequest): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/access-list/batch", request);
    await this.refreshConfig();
    return result;
  }

  /**
   * Reorder rules in an access-list
   */
  async reorderRules(
    identifier: string,
    listType: string,
    rules: Array<{ old_number: number; new_number: number; rule_data: AccessListRule }>
  ): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/access-list/reorder", {
      identifier,
      list_type: listType,
      rules,
    });
    await this.refreshConfig();
    return result;
  }

  /**
   * Helper: Create a new access-list with first rule
   */
  async createAccessList(
    identifier: string,
    listType: string,
    description: string | null,
    rule: Partial<AccessListRule>
  ): Promise<VyOSResponse> {
    const operations: AccessListBatchOperation[] = [];

    // Create access-list
    operations.push({ op: listType === "ipv4" ? "set_access_list" : "set_access_list6" });

    // Add description if provided
    if (description) {
      operations.push({
        op: listType === "ipv4" ? "set_access_list_description" : "set_access_list6_description",
        value: description
      });
    }

    // Create rule
    operations.push({ op: listType === "ipv4" ? "set_rule" : "set_rule6" });

    // Set rule action
    operations.push({
      op: listType === "ipv4" ? "set_rule_action" : "set_rule6_action",
      value: rule.action || "permit"
    });

    // Set rule description
    if (rule.description) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_description" : "set_rule6_description",
        value: rule.description
      });
    }

    // Set source
    if (listType === "ipv4") {
      // IPv4 source handling
      if (rule.source_type === "any") {
        operations.push({
          op: "set_rule_source_any"
        });
      } else if (rule.source_type === "host" && rule.source_address) {
        operations.push({
          op: "set_rule_source_host",
          value: rule.source_address
        });
      } else if (rule.source_type === "inverse-mask" && rule.source_address && rule.source_mask) {
        operations.push({
          op: "set_rule_source_inverse_mask",
          value: rule.source_address,
          value2: rule.source_mask
        });
      } else if (rule.source_type === "network" && rule.source_address && rule.source_mask) {
        operations.push({
          op: "set_rule_source_network",
          value: rule.source_address,
          value2: rule.source_mask
        });
      }
    } else {
      // IPv6 source handling - any can coexist with network or exact-match
      // exact-match and network are mutually exclusive
      if (rule.source_type === "any") {
        operations.push({
          op: "set_rule6_source_any"
        });
      }

      if (rule.source_address) {
        // Network can coexist with "any"
        operations.push({
          op: "set_rule6_source_network",
          value: rule.source_address
        });
      }

      if (rule.source_exact_match) {
        operations.push({
          op: "set_rule6_source_exact_match"
        });
      }
    }

    // Set destination (IPv4 only - IPv6 access-lists don't have destination)
    if (listType === "ipv4") {
      if (rule.destination_type === "any") {
        operations.push({
          op: "set_rule_destination_any"
        });
      } else if (rule.destination_type === "host" && rule.destination_address) {
        operations.push({
          op: "set_rule_destination_host",
          value: rule.destination_address
        });
      } else if (rule.destination_type === "inverse-mask" && rule.destination_address && rule.destination_mask) {
        operations.push({
          op: "set_rule_destination_inverse_mask",
          value: rule.destination_address,
          value2: rule.destination_mask
        });
      } else if (rule.destination_type === "network" && rule.destination_address && rule.destination_mask) {
        operations.push({
          op: "set_rule_destination_network",
          value: rule.destination_address,
          value2: rule.destination_mask
        });
      }
    }

    return this.batchConfigure({
      identifier,
      list_type: listType,
      rule_number: rule.rule_number,
      operations,
    });
  }

  /**
   * Helper: Update access-list description
   */
  async updateAccessListDescription(
    identifier: string,
    listType: string,
    description: string | null
  ): Promise<VyOSResponse> {
    const operations: AccessListBatchOperation[] = [];

    if (description) {
      operations.push({
        op: listType === "ipv4" ? "set_access_list_description" : "set_access_list6_description",
        value: description
      });
    } else {
      operations.push({
        op: listType === "ipv4" ? "delete_access_list_description" : "delete_access_list6_description"
      });
    }

    return this.batchConfigure({
      identifier,
      list_type: listType,
      operations,
    });
  }

  /**
   * Helper: Delete an access-list
   */
  async deleteAccessList(identifier: string, listType: string): Promise<VyOSResponse> {
    const operations: AccessListBatchOperation[] = [];
    operations.push({
      op: listType === "ipv4" ? "delete_access_list" : "delete_access_list6"
    });

    return this.batchConfigure({
      identifier,
      list_type: listType,
      operations,
    });
  }

  /**
   * Helper: Add a rule to an existing access-list
   */
  async addRule(
    identifier: string,
    listType: string,
    rule: Partial<AccessListRule>
  ): Promise<VyOSResponse> {
    const operations: AccessListBatchOperation[] = [];

    // Create rule
    operations.push({ op: listType === "ipv4" ? "set_rule" : "set_rule6" });

    // Set rule action
    operations.push({
      op: listType === "ipv4" ? "set_rule_action" : "set_rule6_action",
      value: rule.action || "permit"
    });

    // Set rule description
    if (rule.description) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_description" : "set_rule6_description",
        value: rule.description
      });
    }

    // Set source
    if (listType === "ipv4") {
      // IPv4 source handling
      if (rule.source_type === "any") {
        operations.push({
          op: "set_rule_source_any"
        });
      } else if (rule.source_type === "host" && rule.source_address) {
        operations.push({
          op: "set_rule_source_host",
          value: rule.source_address
        });
      } else if (rule.source_type === "inverse-mask" && rule.source_address && rule.source_mask) {
        operations.push({
          op: "set_rule_source_inverse_mask",
          value: rule.source_address,
          value2: rule.source_mask
        });
      } else if (rule.source_type === "network" && rule.source_address && rule.source_mask) {
        operations.push({
          op: "set_rule_source_network",
          value: rule.source_address,
          value2: rule.source_mask
        });
      }
    } else {
      // IPv6 source handling - any can coexist with network or exact-match
      // exact-match and network are mutually exclusive
      if (rule.source_type === "any") {
        operations.push({
          op: "set_rule6_source_any"
        });
      }

      if (rule.source_address) {
        // Network can coexist with "any"
        operations.push({
          op: "set_rule6_source_network",
          value: rule.source_address
        });
      }

      if (rule.source_exact_match) {
        operations.push({
          op: "set_rule6_source_exact_match"
        });
      }
    }

    // Set destination (IPv4 only - IPv6 access-lists don't have destination)
    if (listType === "ipv4") {
      if (rule.destination_type === "any") {
        operations.push({
          op: "set_rule_destination_any"
        });
      } else if (rule.destination_type === "host" && rule.destination_address) {
        operations.push({
          op: "set_rule_destination_host",
          value: rule.destination_address
        });
      } else if (rule.destination_type === "inverse-mask" && rule.destination_address && rule.destination_mask) {
        operations.push({
          op: "set_rule_destination_inverse_mask",
          value: rule.destination_address,
          value2: rule.destination_mask
        });
      } else if (rule.destination_type === "network" && rule.destination_address && rule.destination_mask) {
        operations.push({
          op: "set_rule_destination_network",
          value: rule.destination_address,
          value2: rule.destination_mask
        });
      }
    }

    return this.batchConfigure({
      identifier,
      list_type: listType,
      rule_number: rule.rule_number,
      operations,
    });
  }

  /**
   * Helper: Update an existing rule
   */
  async updateRule(
    identifier: string,
    listType: string,
    ruleNumber: number,
    rule: Partial<AccessListRule>
  ): Promise<unknown> {
    const operations: AccessListBatchOperation[] = [];

    // Delete existing source first
    operations.push({
      op: listType === "ipv4" ? "delete_rule_source" : "delete_rule6_source"
    });

    // Delete existing destination (IPv4 only - IPv6 doesn't have destination)
    if (listType === "ipv4") {
      operations.push({
        op: "delete_rule_destination"
      });
    }

    // Update action
    if (rule.action) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_action" : "set_rule6_action",
        value: rule.action
      });
    }

    // Update description
    if (rule.description !== undefined) {
      if (rule.description) {
        operations.push({
          op: listType === "ipv4" ? "set_rule_description" : "set_rule6_description",
          value: rule.description
        });
      } else {
        operations.push({
          op: listType === "ipv4" ? "delete_rule_description" : "delete_rule6_description"
        });
      }
    }

    // Set new source
    if (listType === "ipv4") {
      // IPv4 source handling
      if (rule.source_type === "any") {
        operations.push({
          op: "set_rule_source_any"
        });
      } else if (rule.source_type === "host" && rule.source_address) {
        operations.push({
          op: "set_rule_source_host",
          value: rule.source_address
        });
      } else if (rule.source_type === "inverse-mask" && rule.source_address && rule.source_mask) {
        operations.push({
          op: "set_rule_source_inverse_mask",
          value: rule.source_address,
          value2: rule.source_mask
        });
      } else if (rule.source_type === "network" && rule.source_address && rule.source_mask) {
        operations.push({
          op: "set_rule_source_network",
          value: rule.source_address,
          value2: rule.source_mask
        });
      }
    } else {
      // IPv6 source handling - any can coexist with network or exact-match
      // exact-match and network are mutually exclusive
      if (rule.source_type === "any") {
        operations.push({
          op: "set_rule6_source_any"
        });
      }

      if (rule.source_address) {
        // Network can coexist with "any"
        operations.push({
          op: "set_rule6_source_network",
          value: rule.source_address
        });
      }

      if (rule.source_exact_match) {
        operations.push({
          op: "set_rule6_source_exact_match"
        });
      }
    }

    // Set new destination (IPv4 only - IPv6 access-lists don't have destination)
    if (listType === "ipv4") {
      if (rule.destination_type === "any") {
        operations.push({
          op: "set_rule_destination_any"
        });
      } else if (rule.destination_type === "host" && rule.destination_address) {
        operations.push({
          op: "set_rule_destination_host",
          value: rule.destination_address
        });
      } else if (rule.destination_type === "inverse-mask" && rule.destination_address && rule.destination_mask) {
        operations.push({
          op: "set_rule_destination_inverse_mask",
          value: rule.destination_address,
          value2: rule.destination_mask
        });
      } else if (rule.destination_type === "network" && rule.destination_address && rule.destination_mask) {
        operations.push({
          op: "set_rule_destination_network",
          value: rule.destination_address,
          value2: rule.destination_mask
        });
      }
    }

    return this.batchConfigure({
      identifier,
      list_type: listType,
      rule_number: ruleNumber,
      operations,
    });
  }

  /**
   * Helper: Delete a rule
   */
  async deleteRule(
    identifier: string,
    listType: string,
    ruleNumber: number
  ): Promise<unknown> {
    const operations: AccessListBatchOperation[] = [];
    operations.push({
      op: listType === "ipv4" ? "delete_rule" : "delete_rule6"
    });

    return this.batchConfigure({
      identifier,
      list_type: listType,
      rule_number: ruleNumber,
      operations,
    });
  }
}

export const accessListService = new AccessListService();

File: backend/routers/access_list/access_list.py
"""
Access List Router

API endpoints for managing VyOS access-list configuration.
Supports both IPv4 (access-list) and IPv6 (access-list6) with version-aware configuration.
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders import AccessListBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
from utils.router_helpers import load_vyos_capabilities
import inspect

router = APIRouter(prefix="/vyos/access-list", tags=["access-list"])

# Stub functions for backwards compatibility with app.py
def set_device_registry(registry):
    """Legacy function - no longer used."""
    pass


def set_configured_device_name(name):
    """Legacy function - no longer used."""
    pass


# ============================================================================
# Pydantic Models
# ============================================================================


class AccessListRule(BaseModel):
    """Model for an access-list rule."""
    rule_number: int
    action: str  # permit or deny
    description: Optional[str] = None
    source_type: Optional[str] = None  # any, host, inverse-mask, network
    source_address: Optional[str] = None
    source_mask: Optional[str] = None
    source_exact_match: Optional[bool] = None  # IPv6 exact-match flag
    destination_type: Optional[str] = None  # any, host, inverse-mask, network
    destination_address: Optional[str] = None
    destination_mask: Optional[str] = None


class AccessList(BaseModel):
    """Model for an access-list."""
    number: str  # For IPv4, name for IPv6
    description: Optional[str] = None
    rules: List[AccessListRule] = []
    list_type: str  # ipv4 or ipv6


class AccessListConfigResponse(BaseModel):
    """Response containing all access-list configuration."""
    ipv4_lists: List[AccessList] = []
    ipv6_lists: List[AccessList] = []
    total_ipv4: int = 0
    total_ipv6: int = 0


class AccessListBatchOperation(BaseModel):
    """Single operation in a batch request."""
    op: str = Field(..., description="Operation name")
    value: Optional[str] = Field(None, description="Operation value")
    value2: Optional[str] = Field(None, description="Second value (for inverse-mask, network)")


class AccessListBatchRequest(BaseModel):
    """Model for batch configuration."""
    identifier: str = Field(..., description="Access list number (IPv4) or name (IPv6)")
    list_type: str = Field(..., description="ipv4 or ipv6")
    rule_number: Optional[int] = Field(None, description="Rule number for rule operations")
    operations: List[AccessListBatchOperation]


class ReorderRuleItem(BaseModel):
    """Model for a single rule in reorder request."""
    old_number: int
    new_number: int
    rule_data: AccessListRule


class ReorderAccessListRequest(BaseModel):
    """Model for reordering rules in an access-list."""
    identifier: str = Field(..., description="Access list number (IPv4) or name (IPv6)")
    list_type: str = Field(..., description="ipv4 or ipv6")
    rules: List[ReorderRuleItem]


class VyOSResponse(BaseModel):
    """Standard response from VyOS operations."""
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_access_list_capabilities(request: Request):
    """
    Get feature capabilities based on device VyOS version.

    Returns feature flags indicating which operations are supported.
    Allows frontends to conditionally enable/disable features.
    """
    return await load_vyos_capabilities(
        request=request,
        feature_group=FeatureGroup.ACCESS_LIST,
        builder_cls=AccessListBatchBuilder,
    )


# ============================================================================
# Endpoint 2: Config (Generalized Data)
# ============================================================================


@router.get("/config", response_model=AccessListConfigResponse)
async def get_access_list_config(http_request: Request, refresh: bool = False):
    """
    Get all access-list configurations from VyOS in a generalized format.

    Args:
        refresh: If True, force refresh from VyOS. If False, use cache.

    Returns:
        Generalized configuration data optimized for frontend consumption
    """
    # Check RBAC permission
    await require_read_permission(http_request, FeatureGroup.ACCESS_LIST)

    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)
        
        ipv4_lists = []
        ipv6_lists = []
        
        # Parse IPv4 access-lists
        policy_config = full_config.get("policy", {})
        access_lists = policy_config.get("access-list", {})
        
        for number, list_config in access_lists.items():
            rules = []
            rules_config = list_config.get("rule", {})
            
            for rule_num, rule_config in rules_config.items():
                # Parse source
                source_type = None
                source_address = None
                source_mask = None
                source_data = rule_config.get("source", {})

                if isinstance(source_data, dict):
                    if "any" in source_data:
                        source_type = "any"
                    elif "host" in source_data:
                        source_type = "host"
                        host_val = source_data.get("host")
                        source_address = host_val if isinstance(host_val, str) else None
                    elif "inverse-mask" in source_data and "network" in source_data:
                        # Inverse-mask format: {'inverse-mask': '0.0.0.255', 'network': '192.168.80.0'}
                        source_type = "inverse-mask"
                        source_address = source_data.get("network")
                        source_mask = source_data.get("inverse-mask")
                    elif "network" in source_data:
                        source_type = "network"
                        net_data = source_data.get("network", {})
                        if isinstance(net_data, str):
                            # Direct value (e.g., IPv6 CIDR)
                            source_address = net_data
                        elif isinstance(net_data, dict):
                            # Nested structure: address -> mask -> {}
                            for key, value in net_data.items():
                                source_address = key
                                if isinstance(value, dict) and value:
                                    source_mask = list(value.keys())[0]
                                break  # Only take first entry

                # Parse destination
                dest_type = None
                dest_address = None
                dest_mask = None
                dest_data = rule_config.get("destination", {})

                if isinstance(dest_data, dict):
                    if "any" in dest_data:
                        dest_type = "any"
                    elif "host" in dest_data:
                        dest_type = "host"
                        host_val = dest_data.get("host")
                        dest_address = host_val if isinstance(host_val, str) else None
                    elif "inverse-mask" in dest_data and "network" in dest_data:
                        # Inverse-mask format: {'inverse-mask': '0.0.0.255', 'network': '192.168.80.0'}
                        dest_type = "inverse-mask"
                        dest_address = dest_data.get("network")
                        dest_mask = dest_data.get("inverse-mask")
                    elif "network" in dest_data:
                        dest_type = "network"
                        net_data = dest_data.get("network", {})
                        if isinstance(net_data, str):
                            # Direct value (e.g., IPv6 CIDR)
                            dest_address = net_data
                        elif isinstance(net_data, dict):
                            # Nested structure: address -> mask -> {}
                            for key, value in net_data.items():
                                dest_address = key
                                if isinstance(value, dict) and value:
                                    dest_mask = list(value.keys())[0]
                                break  # Only take first entry
                
                rule = AccessListRule(
                    rule_number=int(rule_num),
                    action=rule_config.get("action", "permit"),
                    description=rule_config.get("description"),
                    source_type=source_type,
                    source_address=source_address,
                    source_mask=source_mask,
                    destination_type=dest_type,
                    destination_address=dest_address,
                    destination_mask=dest_mask,
                )
                rules.append(rule)
            
            # Sort rules by rule number
            rules.sort(key=lambda r: r.rule_number)
            
            access_list = AccessList(
                number=number,
                description=list_config.get("description"),
                rules=rules,
                list_type="ipv4",
            )
            ipv4_lists.append(access_list)
        
        # Parse IPv6 access-lists
        access_lists6 = policy_config.get("access-list6", {})
        
        for name, list_config in access_lists6.items():
            rules = []
            rules_config = list_config.get("rule", {})
            
            for rule_num, rule_config in rules_config.items():
                # Parse source (IPv6 supports any, exact-match, and network)
                # Note: any can coexist with exact-match OR network
                # But exact-match and network are mutually exclusive
                source_type = None
                source_address = None
                source_exact_match = False
                source_data = rule_config.get("source", {})

                if isinstance(source_data, dict):
                    # Check for "any" flag
                    if "any" in source_data:
                        source_type = "any"

                    # Check for "exact-match" flag (can coexist with any)
                    if "exact-match" in source_data:
                        source_exact_match = True

                    # Check for "network" (can coexist with any, but not exact-match)
                    if "network" in source_data:
                        net_val = source_data.get("network")
                        source_address = net_val if isinstance(net_val, str) else None
                        # If we have network but no "any", set source_type to "network"
                        if not source_type:
                            source_type = "network"

                # NOTE: IPv6 access-lists do NOT have destination fields
                # They only match on source

                rule = AccessListRule(
                    rule_number=int(rule_num),
                    action=rule_config.get("action", "permit"),
                    description=rule_config.get("description"),
                    source_type=source_type,
                    source_address=source_address,
                    source_exact_match=source_exact_match,  # Store exact-match flag
                    destination_type=None,  # IPv6 doesn't have destination
                    destination_address=None,  # IPv6 doesn't have destination
                )
                rules.append(rule)
            
            # Sort rules by rule number
            rules.sort(key=lambda r: r.rule_number)
            
            access_list = AccessList(
                number=name,
                description=list_config.get("description"),
                rules=rules,
                list_type="ipv6",
            )
            ipv6_lists.append(access_list)
        
        # Sort access lists by number/name
        ipv4_lists.sort(key=lambda al: int(al.number) if al.number.isdigit() else 0)
        ipv6_lists.sort(key=lambda al: al.number)
        
        return AccessListConfigResponse(
            ipv4_lists=ipv4_lists,
            ipv6_lists=ipv6_lists,
            total_ipv4=len(ipv4_lists),
            total_ipv6=len(ipv6_lists),
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch")
async def access_list_batch_configure(http_request: Request, body: AccessListBatchRequest):
    """
    Execute a batch of configuration operations.

    Allows multiple changes in a single VyOS commit for efficiency.
    """
    # Check RBAC permission
    await require_write_permission(http_request, FeatureGroup.ACCESS_LIST)

    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = AccessListBatchBuilder(version=version)
        
        # Process operations using inspect for dynamic method calls
        for operation in body.operations:
            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = list(sig.parameters.keys())

            # Build arguments dynamically
            args = []

            # Add identifier (number for IPv4, name for IPv6)
            if "number" in params or "name" in params:
                args.append(body.identifier)

            # Add rule number if present in method signature
            if "rule" in params and body.rule_number is not None:
                args.append(str(body.rule_number))

            # Add operation values
            if operation.value and len(params) > len(args):
                args.append(operation.value)

            if operation.value2 and len(params) > len(args):
                args.append(operation.value2)

            method(*args)

        # Execute batch
        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "Configuration updated"},
            error=response.error if response.error else None
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# Additional Helper Endpoints
# ============================================================================
# All operations use the batch endpoint above
# No direct DELETE endpoints - use /batch for all operations
# ============================================================================


# ============================================================================
# Reorder Endpoint
# ============================================================================


@router.post("/reorder")
async def reorder_access_list_rules(http_request: Request, body: ReorderAccessListRequest):
    """
    Reorder rules in an access-list.

    This endpoint deletes all existing rules and recreates them with new numbers
    in a single VyOS commit for atomicity.
    """
    # Check RBAC permission
    await require_write_permission(http_request, FeatureGroup.ACCESS_LIST)

    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = AccessListBatchBuilder(version=version)
        
        # Step 1: Delete all rules in reverse order
        rules_to_delete = sorted(
            [item.old_number for item in body.rules],
            reverse=True
        )
        
        for old_number in rules_to_delete:
            if body.list_type == "ipv4":
                builder.delete_rule(body.identifier, str(old_number))
            else:  # ipv6
                builder.delete_rule6(body.identifier, str(old_number))
        
        # Step 2: Recreate rules with new numbers
        for item in body.rules:
            rule = item.rule_data
            rule_str = str(item.new_number)
            
            if body.list_type == "ipv4":
                # Create rule
                builder.set_rule(body.identifier, rule_str)
                
                # Set action
                builder.set_rule_action(body.identifier, rule_str, rule.action)
                
                # Set description
                if rule.description:
                    builder.set_rule_description(
                        body.identifier, rule_str, rule.description
                    )
                
                # Set source
                if rule.source_type == "any":
                    builder.set_rule_source_any(body.identifier, rule_str)
                elif rule.source_type == "host" and rule.source_address:
                    builder.set_rule_source_host(
                        body.identifier, rule_str, rule.source_address
                    )
                elif rule.source_type == "inverse-mask" and rule.source_address and rule.source_mask:
                    builder.set_rule_source_inverse_mask(
                        body.identifier, rule_str, rule.source_address, rule.source_mask
                    )
                elif rule.source_type == "network" and rule.source_address and rule.source_mask:
                    builder.set_rule_source_network(
                        body.identifier, rule_str, rule.source_address, rule.source_mask
                    )
                
                # Set destination
                if rule.destination_type == "any":
                    builder.set_rule_destination_any(body.identifier, rule_str)
                elif rule.destination_type == "host" and rule.destination_address:
                    builder.set_rule_destination_host(
                        body.identifier, rule_str, rule.destination_address
                    )
                elif rule.destination_type == "inverse-mask" and rule.destination_address and rule.destination_mask:
                    builder.set_rule_destination_inverse_mask(
                        body.identifier, rule_str, rule.destination_address, rule.destination_mask
                    )
                elif rule.destination_type == "network" and rule.destination_address and rule.destination_mask:
                    builder.set_rule_destination_network(
                        body.identifier, rule_str, rule.destination_address, rule.destination_mask
                    )
            
            else:  # ipv6
                # Create rule
                builder.set_rule6(body.identifier, rule_str)
                
                # Set action
                builder.set_rule6_action(body.identifier, rule_str, rule.action)
                
                # Set description
                if rule.description:
                    builder.set_rule6_description(
                        body.identifier, rule_str, rule.description
                    )
                
                # Set source
                if rule.source_type == "any":
                    builder.set_rule6_source_any(body.identifier, rule_str)
                elif rule.source_type == "network" and rule.source_address:
                    builder.set_rule6_source_network(
                        body.identifier, rule_str, rule.source_address
                    )

                # NOTE: IPv6 access-lists do NOT have destination fields
        
        # Execute batch
        response = service.execute_batch(builder)
        
        return VyOSResponse(
            success=response.status == 200,
            data={"message": f"Rules reordered in access-list {body.identifier}"},
            error=response.error if response.error else None
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
